src/esp/bindings_js/topdown.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * TopDownMap class
+ */
+// eslint-disable-next-line no-unused-vars
+class TopDownMap {
+  // PUBLIC methods.
+
+  /**
+   * Create top-down map.
+   * @param {PathFinder} pathFinder - pathFinder used to create map
+   * @param {HTMLCanvasElement} canvas - canvas element for drawing map
+   */
+  constructor(pathFinder, canvas) {
+    this.pathFinder = pathFinder;
+    this.canvas = canvas;
+    this.ctx = canvas.getContext("2d");
+    this.imageData = this.createMap();
+  }
+
+  /**
+   * Draw the topdown map to its canvas.
+   */
+  draw() {
+    this.ctx.putImageData(this.imageData, 0, 0);
+  }
+
+  // PRIVATE methods.
+
+  createMap() {
+    let ctx = this.ctx;
+    let canvas = this.canvas;
+    let pathFinder = this.pathFinder;
+    let imageData = ctx.createImageData(canvas.width, canvas.height);
+    let bounds = pathFinder.bounds;
+    let width = bounds.max[0] - bounds.min[0];
+    let height = bounds.max[2] - bounds.min[2];
+    let scale = 1.0;
+    let heightInPixels = 0;
+    let widthInPixels = 0;
+    let widthComponent = 0;
+    let heightComponent = 2;
+
+    // Best-Fit: orient and scale for tightest fit
+    if (
+      (width > height && canvas.width < canvas.height) ||
+      (width < height && canvas.width > canvas.height)
+    ) {
+      widthComponent = 2;
+      heightComponent = 0;
+      let tmp = width;
+      width = height;
+      height = tmp;
+    }
+    if (height / width > canvas.height / canvas.width) {
+      // Fit height
+      scale = canvas.height / height;
+      heightInPixels = canvas.height;
+      widthInPixels = Math.round(width * scale - 0.5);
+    } else {
+      // Fit width
+      scale = canvas.width / width;
+      widthInPixels = canvas.width;
+      heightInPixels = Math.round(height * scale - 0.5);
+    }
+
+    // Create map by painting navigable locations white
+    let i, j;
+    let stride = 4 * canvas.width;
+    let increment = 1 / scale;
+    let y = bounds.min[heightComponent] + increment / 2;
+    for (i = 0; i < heightInPixels; i++) {
+      let x = bounds.min[widthComponent] + increment / 2;
+      for (j = 0; j < widthInPixels; j++) {
+        let point = [0, 0, 0];
+        point[widthComponent] = x;
+        point[heightComponent] = y;
+        let isNavigable = pathFinder.isNavigable(point, 0.5);
+        if (isNavigable) {
+          imageData.data[i * stride + j * 4] = 0xff;
+          imageData.data[i * stride + j * 4 + 1] = 0xff;
+          imageData.data[i * stride + j * 4 + 2] = 0xff;
+          imageData.data[i * stride + j * 4 + 3] = 0xff;
+        }
+        x += increment;
+      }
+      y += increment;
+    }
+    return imageData;
+  }
+}
```

Comment:
Totally random comment: wouldn't this exact functionality be useful also on the C++/Python side (or alternatively, isn't there one already)? In that case maybe have it implemented in C++ and push only the final binary image to JS?

Reply:
There is a python implementation but not a C++ implementation. I wanted to write a C++ implementation but couldn't come up with an interface I was happy with. I would need to pass in the sampling distance, orientation, and color (or return a 2D vector of booleans). I would still need to js or python specific code for doing the final render. However, a C++ implementation would be way more efficient. I will revisit implementing in C++ in a future change once I can figure out an API I'm happy with.